### PR TITLE
fix: contact names should not allow duplication

### PR DESCRIPTION
**Pull request by Matt561**
## **Description**
This fix prevents users from having duplicate names in their contact list. A warning banner is now displayed in the contact list when duplicates are found. All duplicate contacts will also have a warning icon next to them to help users better identify duplicates.

#### What is considered a duplicate contact?
A duplicate is a contact with a different address but same name as another contact or account.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/28249?quickstart=1)

## **Related issues**

Fixes: 
- https://github.com/MetaMask/metamask-extension/issues/26621
- https://github.com/MetaMask/metamask-extension/issues/26696

#### Adding a duplicate contact
1. Go to Settings > Contacts page
2. Attempt to add a new contact using an existing contact or account name
3. An error message will appear underneath the `username` input and the submit button will be disabled.

#### Viewing duplicate contact warning banner
1. In your local environment, go to file `ui/components/app/contact-list/contact-list.component.js`
2. On line `138` (`{hasDuplicateContacts ? this.renderDuplicateContactWarning() : null}` change `hasDuplicateContacts` to `true` to display the banner.

### **Before**

https://github.com/user-attachments/assets/a7813050-e906-4d42-ac10-db05401d06d3

### **After**

https://github.com/user-attachments/assets/70e16e99-60ea-4404-9ea7-6f776f15d00e

Screenshot 2024-11-15 at 1 59 21 PM

Screenshot 2024-11-15 at 1 59 28 PM

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by Matt561** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by mirceanis — COMMENTED** [review]
I have some questions before I can approve. can you please clarify?

File: ui/pages/settings/contact-list-tab/add-contact/add-contact.component.js
Comment (Matt561) [reply ui/pages/settings/contact-list-tab/add-contact/add-contact.component.js:110]: I decided to do this for the address book and new name so the collision check was case insensitive. This way we wouldn't allow `John` and `john` as separate contacts.

File: ui/components/app/contact-list/contact-list.component.js
Comment (Matt561) [reply ui/components/app/contact-list/contact-list.component.js:108]: I've updated this to include internal account names.

File: test/data/mock-data.js
Comment (Matt561) [reply test/data/mock-data.js:1]: Sorry about that!

**Review by matthewwalsh0 — DISMISSED** [review]
Reviewed minor confirmations ownership changes .

**Review by vinnyhoward — DISMISSED** [review]
LGTM. Pulled down and tested as well, looking good✅

**Review by mirceanis — DISMISSED** [review]
Looks good!

**Comment by Matt561** [comment]
👋 Hey all, the design team requested that I add a warning icon next to the duplicate contacts for better clarity. Could I get another review when you get a moment please? 😄 

Here are the latest screenshots 

Screenshot 2024-11-15 at 1 59 21 PM

Screenshot 2024-11-15 at 1 59 28 PM

**Review by vinnyhoward — APPROVED** [review]

**Review by matthewwalsh0 — APPROVED** [review]

**Review by mirceanis — APPROVED** [review]
looks good!

**Pull request merged**
